pkg/operator/quorumguardcontroller/quorumguardcontroller.go

```diff
@@ -0,0 +1,293 @@
+
+	if c.etcdQuorumGuard == nil {
+		c.etcdQuorumGuard = resourceread.ReadDeploymentV1OrDie(etcd_assets.MustAsset("etcd/quorumguard-deployment.yaml"))
+	}
+	c.etcdQuorumGuard.Spec.Replicas = &replicaCount
+
+	// if restart occurred, we will apply etcd guard deployment but if it is the same, nothing will happened
+	actual, modified, err := resourceapply.ApplyDeploymentv1(ctx, c.kubeClient.AppsV1(), c.etcdQuorumGuard)
+	if err != nil {
+		klog.Errorf("Failed to verify/apply %s, error %w", EtcdGuardDeploymentName, err)
+		return err
+	}
+
+	// if deployment was modified and is not managed by cvo need to save spec from modified one
+	// cause there are some fields added on creation or apply, for example permissions on volumes
+	// and we want to save them to be able to verify that deployment was not changed
+	// if managed by cvo, delete it
+	if !c.findAndDeleteCVOManagedQuorumGuardDeployment(ctx, actual, recorder) && modified {
+		c.etcdQuorumGuard.Spec = actual.Spec
+		msg := fmt.Sprintf("%s was modified", EtcdGuardDeploymentName)
+		klog.Infof(msg)
+		recorder.Event("ModifiedQuorumGuardDeployment", msg)
+	}
+
+	return nil
+}
+
+// deleteCVOManagedQuorumGuardDeployment delete quorumGuard deployment if it is managed by cvo
+// TODO delete after 4.8
+func (c *QuorumGuardController) findAndDeleteCVOManagedQuorumGuardDeployment(ctx context.Context, quorumGuard *appsv1.Deployment, recorder events.Recorder) bool {
+
+	deleteDeployment := func() {
+		if quorumGuard.ObjectMeta.DeletionTimestamp != nil {
+			return
+		}
+
+		klog.Warningf("Deleting cvo-managed etcd quorum guard pdb")
+		err := c.kubeClient.AppsV1().Deployments(operatorclient.TargetNamespace).Delete(ctx, quorumGuard.Name, metav1.DeleteOptions{})
+		if err != nil && !errors.IsNotFound(err) {
+			klog.Warningf("Failed to delete cvo-managed quorum guard pdb: %v", err)
+			return
+		}
+		recorder.Event("CvoManagedQuorumGuardPdbRemoved", "cvo-managed quorum guard pdb has been removed")
+		return
+	}
+
+	for _, managedField := range quorumGuard.ManagedFields {
+		if managedField.Manager == "cluster-version-operator" {
+			deleteDeployment()
+			return true
+		}
+	}
+	return false
+}
+
+// ensureEtcdGuardPdb if etcd quorum guard PDB doesn't exist or ws changed, apply one
+func (c *QuorumGuardController) ensureEtcdGuardPdb(ctx context.Context, recorder events.Recorder) error {
+
+	// if restart occurred, we will apply PDB but if it is the same, nothing will happened
+	actual, modified, err := resourceapply.ApplyPodDisruptionBudgets(ctx, c.kubeClient.PolicyV1beta1(), pdb)
+	if err != nil {
+		klog.Errorf("Failed to verify/apply %s pdb, error %w", EtcdGuardDeploymentName, err)
+		return err
+	}
+
+	// if PDB was modified and is not managed by cvo log and event it
+	// if managed by cvo, delete it
+	if !c.findAndDeleteCVOManagedQuorumGuardPDB(ctx, actual, recorder) && modified {
```

Comment:
I think we want to make this a little more level so maybe we check `findAndDeleteCVOManagedQuorumGuardPDB` first? If CVO managed resource exists then delete it event and return nil. This will result in another sync where we will see it is missing and apply. I am not sure if we care if CVO managed resource was altered we just want to delete it quickly. this way one resource action per sync, do you think this is worthwhile?

- sync
- check isCVOManaged()
-  if true delete and return

#### delete creates new sync
- sync
- check isCVOManaged()
- if false apply and event if modified.

Reply:
I think that current code allows us to create less api calls and in case of freshly install cluster this part of code is just doing nothing without creating any api call. 
In case of upgrade, the first run will try to update and then delete, i think it is not a big issue. We will not have here loop of retries and race. It will happened only once. As i saw from my observations cluster version is the first one to be update and it stops monitoring etcd in this case. 
In my opinion it is not worth changing